### ISSUE TITLE
Hot Reloading Not Working

### ISSUE THREAD

**Issue by AGreenObject** [gatsbyjs/gatsby#13918]
** Summary**
After developing my first hello-world using `gatsby develop` and viewing localhost:8000, I noticed that hot reloading isn't actually working for me. Changes made in VCS code are not being seen in browser, even upon reload of the page.

The only way to see changes I've made in VCS code to index.js is to restart my CLI and re-run `gatsby develop`

**Steps I took**
Command line:
gatsby new hello-world https://github.com/gatsbyjs/gatsby-starter-hello-world#v2
cd hello-world
gatsby develop (however, after this step, I get:

 info ℹ ｢wdm｣: Compiled successfully. )

The terminal then hangs.

changed content in src/pages/index.js using VCS code.

### Environment (if relevant)

  System:
    OS: Linux 4.4 Ubuntu 18.04.1 LTS (Bionic Beaver)
    CPU: (12) x64 Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
    Shell: 4.4.19 - /bin/bash
  Binaries:
    Node: 12.1.0 - /home/linuxbrew/.linuxbrew/bin/node
    npm: 6.9.0 - /home/linuxbrew/.linuxbrew/bin/npm
  npmPackages:
    gatsby: ^2.4.2 => 2.4.2
  npmGlobalPackages:
    gatsby-cli: 2.5.12

**Comment by LekoArts** [comment]
I can't reproduce this on my machine, please try Node v10 or v11 👍

**Comment by AGreenObject** [comment]
@LekoArts I've tried running the same procedure with both Node v10 and v11 with the exact same result. Is there anything else I should try or do you need more information about something?

**Comment by LekoArts** [comment]
Can you try another operating system or code editor? We'd need to find some sort of reproduction or information on how to reproduce this issue :O

**Comment by sidharthachatterjee** [comment]
@AGreenObject Could you inspect the console in your browser and check if the web socket calls (that enable HMR) are being completed?

Perhaps this has to do with web sockets being blocked in your network stack or something?

**Comment by rori4** [comment]
I can confirm I have the same issue. When I change the title in the gatsby config, the title just disappears. I have to redeploy the project in order for it to run properly and apply the change

**Comment by wardpeet** [comment]
@rori4 we sadly don't do HMR for gatsby-node, gatsby-config or gatsby-browser. You'll need to rerun `gatsby develop` or `gatsby build`

**Comment by AGreenObject** [comment]
> @AGreenObject Could you inspect the console in your browser and check if the web socket calls (that enable HMR) are being completed?
> 
> Perhaps this has to do with web sockets being blocked in your network stack or something?

@sidharthachatterjee I'll check this out when I get home this evening.

**Comment by AGreenObject** [comment]
@sidharthachatterjee looks like the web socket isn't being blocked.

![image](https://user-images.githubusercontent.com/9794355/59163870-e1a21200-8acb-11e9-9c79-f9b500aa450d.png)

**Comment by AGreenObject** [comment]
> Can you try another operating system or code editor? We'd need to find some sort of reproduction or information on how to reproduce this issue :O

@LekoArts I've tried another operating system, Windows 10 Home, 10.0.0.17763 Build 17763, using powershell--and I've gotten hot reloading to work. I'm not exactly sure what the issue is/was, but I can tell you that I was using Ubuntu on the WSL (Windows Subsystem for Linux), and I assume the issue that I was having must have something to do with this.

**Comment by wardpeet** [comment]
can you see if anything is being emitted through the websocket by watching eventstream. Could you also try to update [`fs.inotify.max_user_watches`](https://github.com/gatsbyjs/gatsby/issues/11406#issuecomment-458769756) might be a silent error on your end.

**Comment by interglobalmedia** [comment]
Hi, I am having problems with hot reloading as well. Have been on and off, but today was really bad. I updated all my gatsby related plugins in case if there were some straggling version 1 or border version 2 straggling there (started with the gatsby default starter which I downloaded from the repo), and this is what I am working with now:

```
System:
    OS: macOS 10.14.4
    CPU: (8) x64 Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 11.14.0 - ~/.nvm/versions/node/v11.14.0/bin/node
    npm: 6.7.0 - ~/.nvm/versions/node/v11.14.0/bin/npm
  Languages:
    Python: 2.7.15 - /usr/local/bin/python
  Browsers:
    Chrome: 75.0.3770.90
  npmPackages:
    gatsby: ^2.9.4 => 2.9.4
    gatsby-image: ^2.1.4 => 2.1.4
    gatsby-plugin-emotion: ^4.0.7 => 4.0.7
    gatsby-plugin-feed: ^2.2.3 => 2.2.3
    gatsby-plugin-google-analytics: ^2.0.21 => 2.0.21
    gatsby-plugin-manifest: ^2.1.1 => 2.1.1
    gatsby-plugin-offline: ^2.1.3 => 2.1.3
    gatsby-plugin-react-helmet: ^3.0.12 => 3.0.12
    gatsby-plugin-sass: ^2.0.11 => 2.0.11
    gatsby-plugin-sharp: ^2.1.5 => 2.1.5
    gatsby-plugin-sitemap: ^2.1.0 => 2.1.0
    gatsby-plugin-typography: ^2.2.13 => 2.2.13
    gatsby-remark-copy-linked-files: ^2.0.13 => 2.0.13
    gatsby-remark-images: ^3.0.16 => 3.0.16
    gatsby-remark-prismjs: ^3.2.11 => 3.2.11
    gatsby-remark-responsive-iframe: ^2.1.1 => 2.1.1
    gatsby-remark-smartypants: ^2.0.9 => 2.0.9
    gatsby-source-filesystem: ^2.0.39 => 2.0.39
    gatsby-transformer-remark: ^2.4.0 => 2.4.0
    gatsby-transformer-sharp: ^2.1.21 => 2.1.21
  npmGlobalPackages:
    gatsby-cli: 2.5.8
```
Strangely, I also have been getting the following re Workbox in the Chrome DevTools Console:

```
workbox Router is responding to: /
11:16:59.596 workbox-core.dev.js:132 workbox Router is responding to: /component---src-pages-index-js-de75db7165a8435fa31c.js
11:16:59.600 workbox-core.dev.js:132 workbox Router is responding to: /page-data/index/page-data.json
11:16:59.605 workbox-core.dev.js:132 workbox Precaching is responding to: /webpack-runtime-0029c4e5aca2701ef598.js
11:16:59.610 workbox-core.dev.js:132 workbox Precaching is responding to: /app-c3c9e9507e5f82f92060.js
11:16:59.615 workbox-core.dev.js:132 workbox Using NetworkFirst to respond to '/page-data/index/page-data.json'
11:16:59.619 workbox-core.dev.js:132 workbox Precaching is responding to: /styles-fe3bab18b4dc88fb4d66.js
11:16:59.626 workbox-core.dev.js:132 workbox Precaching is responding to: /1-79a8ef40e04afc78e98d.js
11:16:59.632 workbox-core.dev.js:132 workbox Using CacheFirst to respond to '/component---src-pages-index-js-de75db7165a8435fa31c.js'
11:16:59.664 workbox-core.dev.js:132 workbox Router is responding to: https://www.google-analytics.com/analytics.js
11:16:59.673 workbox-core.dev.js:132 workbox Using StaleWhileRevalidate to respond to 'https://www.google-analytics.com/analytics.js'
11:17:00.092 workbox-core.dev.js:132 workbox Precaching is responding to: /manifest.webmanifest
```
Is that supposed to be happening?

**Comment by interglobalmedia** [comment]
Now, when I went into another gatsbyjs v2 project which has been successfully deployed (very recently) after updating OSX (now on version 10.14.5), I am getting the following error in the console which usually goes away on refresh, but now does NOT:

```
Unhandled Rejection (SyntaxError): Unexpected token < in JSON at position 0
XMLHttpRequest.request.onreadystatechange
/Users/mariacam/Development/interglobalmedia/webpack/bootstrap:52
  49 | } else {
  50 | 	// success
  51 | 	try {
> 52 | 		var update = JSON.parse(request.responseText);
  53 | 	} catch (e) {
  54 | 		reject(e);
  55 | 		return;
```

And in the terminal console, I am getting the following:

```
info ℹ ｢wdm｣:
info ℹ ｢wdm｣: Compiled successfully.
Page not found /dev-404-page
Page not found /dev-404-page
Page not found /dev-404-page
Page not found /dev-404-page
Page not found /dev-404-page
Page not found /dev-404-page
Page not found /dev-404-page
```

**Comment by interglobalmedia** [comment]
FYI: I found out in my case that it was a Chrome browser issue. Gatsby works in Firefox and Safari for me but not Chrome. Hope this helps others who might be having similar issues.

**Comment by ken-nah** [comment]
I am also experiencing the same issue. Running on ubuntu 19.04

**Comment by wardpeet** [comment]
Mind sharing your environment by running `gatsby info --clipboard` in your project directory and paste the output here.

It would also be really helpful if you can create a reproduction or share your source code.

**Comment by franklinfarahani** [comment]
I had the same problem where only parts of my project were hot reloading. No error messages or warnings either. It turns out it was because the casing of the component name in my imports was not the same as the actual component folder name.

**Comment by lbittner-pdftron** [comment]
I'm having hot reload issues with certain parts of my project as well. When i make a change to the file i get the `[HMR] bundle rebuilt in Xms` message, but the content on the page does not actually change right away. I have found, however, that if I wait multiple minutes, the page will eventually update.

Running on MacOS 10.14.5.

**Comment by savkevip** [comment]
I have issue to with HR :/

**Comment by mtdenton** [comment]
I ran into this issue, it was solved by creating the project in the WSL2 file structure instead of the Windows file structure.

Original project that couldn't hot reload was created in C:\Users\username\Documents

Project that was able to utilize hot reloading was created in \\wsl$\Ubuntu\home\username (it's all under a "network device" apparently)

Quick tip for this also, you can open file explorer to your current WSL2 directory by running the command "explorer.exe .". It can also be accessed easily in VS Code by installing the VS Code Remote Development plugin, found here: https://code.visualstudio.com/docs/remote/remote-overview

**Comment by RLloyd** [comment]
npm run develop -o

**Comment by gkatzioras** [comment]
> I ran into this issue, it was solved by creating the project in the WSL2 file structure instead of the Windows file structure.
> 
> Original project that couldn't hot reload was created in C:\Users\username\Documents
> 
> Project that was able to utilize hot reloading was created in \wsl$\Ubuntu\home\username (it's all under a "network device" apparently)
> 
> Quick tip for this also, you can open file explorer to your current WSL2 directory by running the command "explorer.exe .". It can also be accessed easily in VS Code by installing the VS Code Remote Development plugin, found here: https://code.visualstudio.com/docs/remote/remote-overview

@mtdenton's advice worked for me to enable HML, but I also had to run "**code .**" inside my WSL project folder. [dont forget the **dot "."** at the end!]

**Comment by NateRadebaugh** [comment]
For what it's worth, I was hitting this issue myself using a new gatsby-theme-blog starter project and resolved the issue by updating my dependencies. The starter repo had out of date dependencies when it was cloned/installed, so updating the version had the fixes I needed!

**Comment by keith-E** [comment]
> I ran into this issue, it was solved by creating the project in the WSL2 file structure instead of the Windows file structure.
> 
> Original project that couldn't hot reload was created in C:\Users\username\Documents
> 
> Project that was able to utilize hot reloading was created in \wsl$\Ubuntu\home\username (it's all under a "network device" apparently)
> 
> Quick tip for this also, you can open file explorer to your current WSL2 directory by running the command "explorer.exe .". It can also be accessed easily in VS Code by installing the VS Code Remote Development plugin, found here: https://code.visualstudio.com/docs/remote/remote-overview

This worked for me also, thanks! The WSL2 terminal was mounting directly to my Windows 10 user folder by default, so I created my Gatsby project there. After recreating my gatsby project (`gatsby new projectname`) in the WSL2/Linux file structure, hot reloading worked.

Also, a huge bonus in my case was that my compilation times (after performing a `gatsby develop`) went down from ~30000ms to ~3500ms.

**Comment by fullspectrumprocessing** [comment]
Are there any solutions for this issue on Mac?? Having the same problem with some of the components not hot reloading. Have to restart the whole server just to get updates.

**Comment by Tenkir** [comment]
Having trouble with hot reloading as well. Works fine if I change an import in my component file, but changing the JSX return does not trigger a hot reload (for example, updating text in an `h1` tag).

**Comment by Tenkir** [comment]
Hot Reloading doesnt seem to work if I nest a component in a subdirectory inside `src/components`. For example, `src/components/header.js` reloads fine, while `src/components/header/index.js` does not.

**Comment by fpigeonjr** [comment]
> 
> 
> I ran into this issue, it was solved by creating the project in the WSL2 file structure instead of the Windows file structure.
> 
> Original project that couldn't hot reload was created in C:\Users\username\Documents
> 
> Project that was able to utilize hot reloading was created in \wsl$\Ubuntu\home\username (it's all under a "network device" apparently)
> 
> Quick tip for this also, you can open file explorer to your current WSL2 directory by running the command "explorer.exe .". It can also be accessed easily in VS Code by installing the VS Code Remote Development plugin, found here: https://code.visualstudio.com/docs/remote/remote-overview

I can confirm that running on the WSL side does fix hot-reloading and the performance is a lot better. My CPU was running hot on the windows side this is the only way to live.

**Comment by facultymatt** [comment]
I see a behavior similar to @Tenkir but with case sensitivity. 

```
filename: src/components/Header/index.js
does not work: import Header from './header/
does work: import Header from './Header/
```

**Comment by gixxerblade** [comment]
> I ran into this issue, it was solved by creating the project in the WSL2 file structure instead of the Windows file structure.
> 
> Original project that couldn't hot reload was created in C:\Users\username\Documents
> 
> Project that was able to utilize hot reloading was created in \wsl$\Ubuntu\home\username (it's all under a "network device" apparently)
> 
> Quick tip for this also, you can open file explorer to your current WSL2 directory by running the command "explorer.exe .". It can also be accessed easily in VS Code by installing the VS Code Remote Development plugin, found here: https://code.visualstudio.com/docs/remote/remote-overview

Amazing! It worked! Thank you

**Comment by wardpeet** [comment]
This might be a HMR bug and not a gastby bug. Is it possible to create a new bug with all the information you can put in.

Does this only happen on windows?

**Comment by ethernal** [comment]
I have same behavior with Chrome (Windows). Enterprise settings in Chrome, some content is blocked etc. please check Firefox if it is unlocked. Had similar issue with Cypress also due to Chrome policy.

**Comment by AAverin** [comment]
I also have issues with hot reload.
Difficult to tell when it started exactly, in `gatsby develop` when I save whole page reloads, indeed, but I kind of expect only one component to update on screen without whole page updating. Am I right to expect that?

Anyways, I also have aliases set up via `gatsby-plugin-alias-imports`. Others mention some import-related issues, so could be related to this also.

**Comment by wispyco** [comment]
Well not sure if this will help everyone, but I had setup WSL 2 through the insiders program, and when I downgraded my ubuntu version to wsl1 hot reload started working again. The command I had to run was `wsl --set-version Ubuntu 1`  more info here about WSL 2 https://docs.microsoft.com/en-us/windows/wsl/wsl2-install

**Comment by Nurou** [comment]
Having the exact same issue @Tenkir

**Comment by seandearnaley** [comment]
pulling my hair over WSL 2 hot reloading, doesn't seem to be detecting changes even though I'm in the linux file structure and webpack hmr seems to be working in the browser... it's like it's not seeing changes when I save then because gatsby develop log shows nothing.

Doesn't seem to be working at all, tried everything.

**Comment by emthorne** [comment]
I had this same issue but solved it by updating my dependencies. 
For reference: https://flaviocopes.com/update-npm-dependencies/

npm install -g npm-check-updates
ncu -u
npm update

This fixed it for me. Hope it helps.

**Comment by mfrench71** [comment]
npm update fixed this for me on Mac.

**Comment by RapidOwl** [comment]
I've got this issue right now. Cloned the Hello World example a couple of hours back and been trying to get HMR working ever since.

The request to `http://localhost:8000/__webpack_hmr`  shows as pending in the Chrome dev tools and then changes to cancelled after a while. The CLI is definitely rebuilding my site and when the rebuild happens, Edge flashes up a localhost request as if it's doing something, but nothing changes.

`npm update` incremented Gatsby's version, but didn't fix the problem.

> System:
> OS: Windows 10 10.0.18363
> CPU: (8) x64 Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
> Binaries:
> Node: 12.15.0 - C:\Program Files\nodejs\node.EXE
> npm: 6.13.4 - C:\Program Files\nodejs\npm.CMD
> Languages:
> Python: 2.7.13
> Browsers:
> Edge: 44.18362.449.0
> npmPackages:
> gatsby: ^2.19.16 => 2.19.16

**Comment by KPS-GitHub** [comment]
Just in case it helps anyone like me, the fix for my little project was to run a 'gatsby build' before running 'gatsby develop'. No idea why this fixed my hot reload, but it's the only thing that's worked for me after a couple days of searching forums and trying a million different solutions.

**Comment by ryancwalsh** [comment]
I'm on a Windows 10 PC running a Laravel Homestead Vagrant Ubuntu VM, where I've been installing the Gatsby tutorial. 

Unfortunately hot reloading has never worked yet.

After changing any code at all, I always need to completely stop and rerun `gatsby develop --host=0.0.0.0`.

My system already says:
```
cat /proc/sys/fs/inotify/max_user_watches
524288
```
so https://github.com/gatsbyjs/gatsby/issues/13918#issuecomment-500709995 didn't help.

[`npm update`](https://github.com/gatsbyjs/gatsby/issues/13918#issuecomment-581133717) didn't help because of https://github.com/gatsbyjs/gatsby/issues/6834#issuecomment-619195994

And running [`gatsby build` before running `gatsby develop`](https://github.com/gatsbyjs/gatsby/issues/13918#issuecomment-592240753) didn't help.

`yarn add react-dom@npm:@hot-loader/react-dom` didn't [help](https://github.com/gatsbyjs/gatsby/issues/11934#issuecomment-596124702).

Any other ideas? Thanks.

**Comment by ryancwalsh** [comment]
Hot reloading seems to work if I open PowerShell as Admin and run `gatsby develop`. I guess I'll just try to use Homestead Vagrant for the Laravel PHP / MySQL parts and then Windows for developing the Gatsby layer.

**Comment by aaronadamsCA** [comment]
The lack of support for hot reloading on WSL2 is a known issue that should eventually be resolved:

https://github.com/microsoft/WSL/issues/4739

TLDR, Windows file system changes are not reported to Linux, therefore hot reloading cannot do its thing.

**Comment by fmaiabatista** [comment]
**WSL 1**
cmd `ver` gives Windows build 10.0.18363.836
PowerShell `wsl -l -v` gives error which means it's v1 according to [this](https://askubuntu.com/questions/1177729/wsl-am-i-running-version-1-or-version-2).

Tried the following without success...
- localhost
- 127.0.0.1
- Files in Windows filesystem (C:\Users\user\Projects\project)
- Files in Linux filesystem (~/Projects/project)

![image](https://user-images.githubusercontent.com/1574224/84403143-c0429d00-abdb-11ea-90bd-cce2527fd8d0.png)

Worth mentioning that this PC has some tight security config that prevents me from accessing some websites such as Playstation's Store and even some news portals. I have no expertise in network security stuff though.

**Comment by dancingmonkey** [comment]
> For what it's worth, I was hitting this issue myself using a new gatsby-theme-blog starter project and resolved the issue by updating my dependencies. The starter repo had out of date dependencies when it was cloned/installed, so updating the version had the fixes I needed!

This worked for me! Thanks! Also finally hot reloading Jekyll sites 💯 %

**Comment by dancingmonkey** [comment]
> > I ran into this issue, it was solved by creating the project in the WSL2 file structure instead of the Windows file structure.
> > Original project that couldn't hot reload was created in C:\Users\username\Documents
> > Project that was able to utilize hot reloading was created in \wsl$\Ubuntu\home\username (it's all under a "network device" apparently)
> > Quick tip for this also, you can open file explorer to your current WSL2 directory by running the command "explorer.exe .". It can also be accessed easily in VS Code by installing the VS Code Remote Development plugin, found here: https://code.visualstudio.com/docs/remote/remote-overview
> 
> This worked for me also, thanks! The WSL2 terminal was mounting directly to my Windows 10 user folder by default, so I created my Gatsby project there. After recreating my gatsby project (`gatsby new projectname`) in the WSL2/Linux file structure, hot reloading worked.
> 
> Also, a huge bonus in my case was that my compilation times (after performing a `gatsby develop`) went down from ~30000ms to ~3500ms.

Thanks so much!! This worked for me!! I quess for now we need to keep the websites inside ubuntu itself on wsl1/wsl2 for hot reloading to work. I had the same issue with Jekyll sites not live reloading, but now it's doing it..happy 👍

**Comment by RilDev** [comment]
Solved by adding a `src/pages/404.js` file.

```
import React from 'react'

const PageNotFound = () => {
  return <>404, page not found!</>
}

export default PageNotFound
```

**Comment by John-Di** [comment]
> Just in case it helps anyone like me, the fix for my little project was to run a 'gatsby build' before running 'gatsby develop'. No idea why this fixed my hot reload, but it's the only thing that's worked for me after a couple days of searching forums and trying a million different solutions.

Just installed Gatsby and cloned the Hello World starter repo with Node version v10.19.0 and Mac Mojave 10.14.6. Can confirm that running 'gatsby build' and then 'gatsby develop' fixed the issue for me (for now). Much appreciated!

**Comment by machineghost** [comment]
This seems really wrong.  Gatsby's `build` command is for *production*: it *should* have nothing to do with the dev server.

**Comment by Neilblaze** [comment]
WSL2 is creating issues on hot-reloading. Don't know why, but I followed [this](https://github.com/gatsbyjs/gatsby/issues/13918#issuecomment-536088178), it still doesn't work!

@Micorsoft should look into this asap!

**Comment by holylander** [comment]
Hi friends!

Im running `gatsby develop` on windows. After trying many things, including wsl, npm clenaup etc, I realized that my antivirus _sophos_ was blocking the webpack hot reload socket `__webpack_hmr`. 

> All I had to do is to disable the antivirus for a while. ( You are free to uninstall it or maybe change to another one ).

hope it works for you ;)

**Comment by emalindahk** [comment]
I had the same problem. What worked for me was to ;

- Delete package-lock.json and node_modules folder
- Run `npm install`

**Comment by machineghost** [comment]
Ah the old "delete `node_modules` and running `npm i`" ... the modern JS equivalent of "blowing on an video game cartridge to make it work" ;-)

And as is (too) often the case ... it worked :-) Thanks!

**Comment by Phillywonka** [comment]
For me `gatsby-plugin-preact` plugin was breaking stuff.

**Comment by tasesmuemils** [comment]
I updated everything in with npm. Before update, everything worked fine, but after the update, hot reload stopped working ([HMR] lines show up in console, but the page content/CSS doesn't update, after manual page refresh - content updates).
What I tried:
Deleted node_modules, .cache, package-lock.jason, npm install - problem still exists
gatsby build and then gatsby develop - reload not working
Tried solutions with @hot-reload/react-dom - reload not working
Using Ubuntu 20.
![Screenshot from 2021-01-26 11-18-06](https://user-images.githubusercontent.com/18176025/105825791-a7e78d00-5fc8-11eb-9d7d-bcac656c5436.png)

**Comment by tasesmuemils** [comment]
> I updated everything in with npm. Before update, everything worked fine, but after the update, hot reload stopped working ([HMR] lines show up in console, but the page content/CSS doesn't update, after manual page refresh - content updates).
> What I tried:
> Deleted node_modules, .cache, package-lock.jason, npm install - problem still exists
> gatsby build and then gatsby develop - reload not working
> Tried solutions with @hot-reload/react-dom - reload not working
> Using Ubuntu 20.
> Chrome dont show any error, Firefox shows this:
> ![Screenshot from 2021-01-26 11-18-06](https://user-images.githubusercontent.com/18176025/105825791-a7e78d00-5fc8-11eb-9d7d-bcac656c5436.png)

Ok, my problem was that I updated react and react-dom to Latest verion (17.0.1). Downgraded back to Wanted (16.14.0). Fixed the problem!

**Comment by logemann** [comment]
Same error here. From one day to another. Dont know if its just one project or all of mine. Whats happening is:

![Bildschirmfoto 2021-01-29 um 13 56 22](https://user-images.githubusercontent.com/346353/106277699-cef8c580-6239-11eb-82a8-f29b86a39b63.png)

See the right EventStream Pane of the chrome dev console. My changes get noticed by HMR because i always get a new event but only sometimes the hash changes. If it does, everything is fine, but most of the time it doesnt change so i dont get an update.

Some things i checked:
* file is updated on fs
* completely cleaned and wiped out caches and node_modules to no avail

Quite clueless here.

**Comment by eth0izzle** [comment]
I was having the same issue. I was importing components with their filename as uppercase (Dashboard via dashboard). Whilst they rendered fine hot-reload was not working. Seems like a bug as on Windows Dashboard and dashboard are the same file and should be treated as such.

**Comment by gajrajgchouhan** [comment]
https://github.com/gatsbyjs/gatsby/issues/13918#issuecomment-679022714

This worked for me !! We need to have a 404 page it seems, you can also see the 404 error in chrome-dev-tools console too.

**Comment by noahgary** [comment]
Doesn't work for me. I have a 404 page and HMR is connected. Doesn't releoad when I save my changes. What's more, is some changes I have don't get loaded at all. I applied an `h-full` class to one of my divs and it just doesn't show up at all. I can edit the class from the console and it works. For some reason, there's a disconnect somewhere.

**Comment by cigzigwon** [comment]
Babel plugin presets are breaking this. Remove them. In my case it babel-preset-gatsby it uses for browserslist but was completely breaking HMR. Fresh install.

**Comment by pbxr250** [comment]
I've found that it's not working in<StaticQuery .... render.... > but working in usual jsx render

**Comment by dko-slapdash** [comment]
@cigzigwon Can you please add more details on your message? How exactly did you solve the problem?

**Comment by lucena-fr4ct1ons** [comment]
If anyone using WSL2 reaches this thread after googling about hot reloading not working, I'd just like to leave my two cents and say that @seandearnaley 's [answer](https://github.com/gatsbyjs/gatsby/issues/13918#issuecomment-570997155) really helped me. Not only it fixed my issue, but it also made the project incredibly faster. I don't know if I was wrong in storing my project on `mnt/c/` but in any case, using the home folder did it.
Thanks, Sean!

**Comment by cigzigwon** [comment]
@dko-slapdash Ummm. Facepalm.

**Comment by saarangsoltani** [comment]
> I had the same problem where only parts of my project were hot reloading. No error messages or warnings either. It turns out it was because the casing of the component name in my imports was not the same as the actual component folder name.

OMG THANK YOU 🙏🏻 🤦🏼‍♂️

**Comment by cigzigwon** [comment]
Guess what? Try using Mapbox SDK V2. I think my HMR is broken again and my guess is that it's all related to babel. Out of the box Gatsby seems to have issues with HMR when you either customize babel or avoid transpiring like with Mapbox. This sucks.

**Comment by cigzigwon** [comment]
So after spending an hour doing a yarn upgrade and patching nothing happened. So I looked a a problem page that didn't load mapbox and what I found is that I have a list of events outside of the page component which is an array of objects with a gatsby <StaticImage/ > tag in it. I just moved that list into the component and it works now! So something similar might be the issue on the mapbox page or that page is going to be an altogether different issue.

**Comment by cigzigwon** [comment]
Oh sweet! cleared the cache rebooted gatsby and mapbox page is fixed! I think this is related to the default-starter being out of date really bad. You have to patch all kindsa stuff if you actually resolve the warning then worker-loader can perform correctly but it had errors until I dumped .cache dir.

**Comment by seshaljain** [comment]
```
gatsby clean
gatsby build
gatsby develop
```
worked for me too

**Comment by EL-MEHDI-ESSAADI** [comment]
Is hot reloading working in windows with Gatsby, or **I have to use Linux**?

**Comment by EL-MEHDI-ESSAADI** [comment]
I **fixed** hot refreshing by **removing** the following option from gatsby-config.js :
``` 
   flags: {
      DEV_SSR: true,
   },

```

**Comment by Kishanjay** [comment]
_Commenting for future reference:_ 

I had a similar problem with an older Gatsby site that I just updated (v4 -> v5).

The issue that I found was that for components which default exported the `React.FC` directly HMR wasn't working. Solution was to store the `React.FC` in a variable first and then export this variable.

e.g. in pages/index.tsx
```
export default () => {
  return (
    <h1>Hello world</h1>
  )
}
```

Change to
```
const IndexPage = () => {
  return (
    <h1>Hello world</h1>
  )
}
export default IndexPage
```

It's strange but seems to fix the issue.
